lib/backend/firestore/firestorebk.go

```diff
@@ -520,33 +539,38 @@ func (b *Backend) CompareAndSwap(ctx context.Context, expected backend.Item, rep
 		return nil, trace.BadParameter("expected and replaceWith keys should match")
 	}
 
-	ref := b.svc.Collection(b.CollectionName).Doc(b.keyToDocumentID(expected.Key))
-	err := b.svc.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error {
-		expectedDocSnap, err := tx.Get(ref)
+	// The operations performed are get, compare values, and conditional update instead of using
+	// a transaction because the firestore api does not return the commit time from RunTransaction.
+	// See https://github.com/googleapis/google-cloud-go/issues/6841.
+	for i := 0; i < 3; i++ {
+		docSnap, err := b.svc.Collection(b.CollectionName).
+			Doc(b.keyToDocumentID(expected.Key)).
+			Get(ctx)
 		if err != nil {
-			return trace.CompareFailed("error or object not found, error: %v", ConvertGRPCError(err))
+			err = ConvertGRPCError(err)
+			if trace.IsNotFound(err) {
+				return nil, trace.CompareFailed(err.Error())
+			}
+			return nil, trace.Wrap(err)
 		}
 
-		existingRecord, err := newRecordFromDoc(expectedDocSnap)
+		r, err := newRecordFromDoc(docSnap)
 		if err != nil {
-			return trace.Wrap(err)
+			return nil, trace.Wrap(err)
 		}
 
-		if !bytes.Equal(existingRecord.Value, expected.Value) {
-			return trace.CompareFailed("expected item value %v does not match actual item value %v", string(expected.Value), existingRecord.Value)
+		if !bytes.Equal(r.Value, expected.Value) {
+			return nil, trace.CompareFailed("expected item value %v does not match actual item value %v", string(expected.Value), r.Value)
 		}
 
-		if err := tx.Set(ref, newRecord(replaceWith, b.clock)); err != nil {
-			return ConvertGRPCError(err)
+		replaceWith.Revision = expected.Revision
+		lease, err = b.ConditionalUpdate(ctx, replaceWith)
+		if err == nil {
+			break
 		}
```

Comment:
I think we should match dynamodb's behavior in `ConditionalUpdate` and `ConditionalDelete`, and return `backend.ErrIncorrectRevision` only if the error we get from the firestore operation is a precondition failed error, otherwise we risk ignoring actual errors caused by network problems or by firestore service problems.